nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java

```diff
@@ -553,69 +552,63 @@ String whereConditionForInstanceUpdate() {
-        for (OptimisticLockKey instance : instances) {
-          batchArgs.add(new Object[] { instance.id, sqlVariants.tuneTimestampForDb(instance.modified) });
-          ids.add(instance.id);
-        }
-        int[] updateStatuses = jdbc
-            .batchUpdate(updateInstanceForExecutionQuery() + " where id = ? and modified = ? and executor_id is null", batchArgs);
-        Iterator<Long> idIt = ids.iterator();
-        for (int status : updateStatuses) {
-          idIt.next();
-          if (status == 0) {
-            idIt.remove();
-            if (ids.isEmpty()) {
-              throw new PollingRaceConditionException("Race condition in polling workflow instances detected. "
-                  + "Multiple pollers using same name (" + executorInfo.getExecutorGroup() + ")");
-            }
-            continue;
-          }
-          if (status != 1 && status != Statement.SUCCESS_NO_INFO) {
-            throw new PollingRaceConditionException("Race condition in polling workflow instances detected. "
-                + "Multiple pollers using same name (" + executorInfo.getExecutorGroup() + ")");
-          }
-        }
+    if (instances.isEmpty()) {
+      return emptyList();
+    }
+    sort(instances);
+    List<Long> ids = new ArrayList<>(instances.size());
+    transaction.execute(transactionStatus -> {
+      if (useBatchUpdate()) {
+        updateNextWorkflowInstancesWithBatchUpdate(instances, ids);
+      } else {
+        updateNextWorkflowInstancesWithMultipleUpdates(instances, ids);
       }
+      return null;
     });
-  }
+    if (ids.isEmpty()) {
+      throw new PollingRaceConditionException("Race condition in polling workflow instances detected. "
+              + "Multiple pollers using same name (" + executorInfo.getExecutorGroup() + ")");
+    }
+    return ids;
+  }
+
+   private void updateNextWorkflowInstancesWithMultipleUpdates(List<OptimisticLockKey> instances, List<Long> ids) {
+     for (OptimisticLockKey instance : instances) {
+       int updated = jdbc.update(updateInstanceForExecutionQuery() + " where id = ? and modified = ? and executor_id is null",
+               instance.id, sqlVariants.tuneTimestampForDb(instance.modified));
+       if (updated == 1) {
+         ids.add(instance.id);
+       }
+     }
+   }
+
+   private void updateNextWorkflowInstancesWithBatchUpdate(List<OptimisticLockKey> instances, List<Long> ids) {
+     List<Object[]> batchArgs = new ArrayList<>(instances.size());
+     for (OptimisticLockKey instance : instances) {
+       batchArgs.add(new Object[] { instance.id, sqlVariants.tuneTimestampForDb(instance.modified) });
+       ids.add(instance.id);
+     }
+     int[] updateStatuses = jdbc
+             .batchUpdate(updateInstanceForExecutionQuery() + " where id = ? and modified = ? and executor_id is null", batchArgs);
+     Iterator<Long> idIt = ids.iterator();
+     for (int status : updateStatuses) {
+       idIt.next();
+       if (status == 1) {
```

Comment:
I guess it should never happen, but what if status > 1? Now it's handled like status == 1, except that it makes to more checks for the value. So this could be status > 0 here?

Reply:
Good point. Rewrote the logic a bit to be shorter and consider values > 1 a failure